server/broadcast.go

```diff
@@ -0,0 +1,214 @@
+		seg.Name = uri // hijack seg.Name to convey the uploaded URI
+	}
+	err = cpl.InsertHLSSegment(vProfile, seg.SeqNo, uri, seg.Duration)
+	if monitor.Enabled {
+		monitor.LogSourceSegmentAppeared(nonce, seg.SeqNo, string(mid), vProfile.Name)
+		glog.V(6).Infof("Appeared segment %d", seg.SeqNo)
+	}
+	if err != nil {
+		glog.Errorf("Error inserting segment %d: %v", seg.SeqNo, err)
+		if monitor.Enabled {
+			monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+		}
+	}
+
+	if sess == nil {
+		return
+	}
+	go func() {
+		// storage the orchestrator prefers
+		if ios := sess.OrchestratorOS; ios != nil {
+			// XXX handle case when orch expects direct upload
+			uri, err := ios.SaveData(name, seg.Data)
+			if err != nil {
+				glog.Error("Error saving segment to OS ", err)
+				if monitor.Enabled {
+					monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+				}
+				return
+			}
+			seg.Name = uri // hijack seg.Name to convey the uploaded URI
+		}
+
+		// send segment to the orchestrator
+		glog.V(common.DEBUG).Infof("Submitting segment %d", seg.SeqNo)
+
+		res, err := SubmitSegment(sess, seg, nonce)
+		if err != nil {
+			if shouldStopStream(err) {
+				glog.Warningf("Stopping current stream due to: %v", err)
+				rtmpStrm.Close()
+			}
+			if shouldStopSession(err) {
+				cxn.needOrch <- struct{}{}
+			}
+			return
+		}
+
+		// download transcoded segments from the transcoder
+		gotErr := false // only send one error msg per segment list
+		errFunc := func(subType, url string, err error) {
+			glog.Errorf("%v error with segment %v: %v (URL: %v)", subType, seg.SeqNo, err, url)
+			if monitor.Enabled && !gotErr {
+				monitor.LogSegmentTranscodeFailed(subType, nonce, seg.SeqNo, err)
+				gotErr = true
+			}
+		}
+
+		segHashes := make([][]byte, len(res.Segments))
+		n := len(res.Segments)
+		SegHashLock := &sync.Mutex{}
+		cond := sync.NewCond(SegHashLock)
+
+		dlFunc := func(url string, i int) {
+			defer func() {
+				cond.L.Lock()
+				n--
+				if n == 0 {
+					cond.Signal()
+				}
+				cond.L.Unlock()
+			}()
+
+			if bos := sess.BroadcasterOS; bos != nil && !drivers.IsOwnExternal(url) {
+				data, err := drivers.GetSegmentData(url)
+				if err != nil {
+					errFunc("Download", url, err)
+					return
+				}
+				name := fmt.Sprintf("%s/%d.ts", sess.Profiles[i].Name, seg.SeqNo)
+				newUrl, err := bos.SaveData(name, data)
+				if err != nil {
+					errFunc("SaveData", url, err)
+					return
+				}
+				url = newUrl
+
+				hash := crypto.Keccak256(data)
+				SegHashLock.Lock()
+				segHashes[i] = hash
+				SegHashLock.Unlock()
+			}
+
+			if monitor.Enabled {
+				monitor.LogTranscodedSegmentAppeared(nonce, seg.SeqNo, sess.Profiles[i].Name)
+			}
+			err = cpl.InsertHLSSegment(&sess.Profiles[i], seg.SeqNo, url, seg.Duration)
+			if err != nil {
+				errFunc("Playlist", url, err)
+				return
+			}
+		}
+
+		for i, v := range res.Segments {
+			go dlFunc(v.Url, i)
+		}
+
+		cond.L.Lock()
+		for n != 0 {
```

Comment:
Curious what your thoughts are on [sync.WaitGroup](https://golang.org/pkg/sync/#WaitGroup)?

```
var wg sync.WaitGroup

for i, v := range res.Segments {
    wg.Add(1)
    go func(i int, url string) {
        defer wg.Done()
        dlFunc(url, i)
    }(i, v.Url)
}

wg.Wait()
```

Reply:
Sure, the reason wait groups are not used here is that I didn't know about them when this was implemented. The refactored code was just copy-pasted from the existing mediaserver code; can fix up the style issues separately.

Reply:
Cool! Didn't know about it either